### DISCUSSION TITLE
Lesson 4: HTTPConnectionPool Error, Can't connect to Ganache (Video Ref:  3:58:45) HELP!!

### DISCUSSION THREAD

**Discussion by abdullahpichoo** [smartcontractkit/full-blockchain-solidity-course-py#1583] [Q&A] [answered]
Hi! I was following along with the tutorial and ran into many issues but I got around most of them. 
**This issue however I cannot find a solution, anywhere.**

Error: HTTPConnectionPool(host='0.0.0.0', port=8545): Max retries exceeded with url
![image_2022-07-15_193646622](https://user-images.githubusercontent.com/98353306/179245722-4df34035-290d-429f-8585-8d8638af8f5f.png)

I got this error where he printed the nonce. _Timestamp: 3:58:45_

- I've tried changing the port of Ganache. It was initially set to 7545 and the network id as 5777. I even changed it to match as in the video but still didn't fix the issue.
- I've restarted VS Code and Ganache but didn't work.
- I've tried turning the firewall off but didn't work. 

Tried searching on Google, but didn't get any solution that I could understand and implement. 

I'm guessing it's related to Ganache cli (which I have no idea how to install and use)

Code:
```Python
from solcx import compile_standard, install_solc
from web3 import Web3

install_solc("0.8.12")

import json

with open("./SimpleStorage.sol", "r") as file:
    simple_storage_file = file.read()

compiled_sol = compile_standard(
    {
        "language": "Solidity",
        "sources": {"SimpleStorage.sol": {"content": simple_storage_file}},
        "settings": {
            "outputSelection": {
                "*": {"*": ["abi", "metadata", "evm.bytecode", "evm.sourceMap"]}
            }
        },
    },
    solc_version="0.8.12",
)

with open("compiled_code.json", "w") as file:
    json.dump(compiled_sol, file)

bytecode = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["evm"][
    "bytecode"
]["object"]

abi = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["abi"]

url = "http://0.0.0.0:8545"
web3 = Web3(Web3.HTTPProvider(url))
chain_id = 1337
chain_address = "0x70DAE3FaD16625670b3149287c42931Eb8049aa6"
chain_privKey = "0x67b4f02d598e44293c5757f181d2e21f850ce4472ccb13ed79e6c301b98d406b"


SimpleStorage = web3.eth.contract(abi=abi, bytecode=bytecode)
nonce = web3.eth.getTransactionCount(chain_address)
print(nonce)

```

**Comment by cromewar** [comment, marked as answer]
Hello @abdullahpichoo 
Thanks for the detailed question, this is really helpful, I have some questions and recommendations for you.
1. Are you using a RPC connected to WSL?
2. use the cahinId: 1337 this might help.
3. Could you please share screenshot of your ganache?

**Reply by abdullahpichoo** [reply]
Hey! Thanks for responding. I actually fixed it somehow.

- So initially the HTTP URL was set to 127.0.0.1, which did not work when I started. So, I changed to 0.0.0.0 in the ganache app settings.
- Installed ganache cli through npm (didn't know I needed node.js for npm) and tried again but it didn't work as well. 
- So I was turning switches at this point, I changed the HTTP URL back to 127.0.0.1 which was my local URL in Ganache's settings (not cli)
- This fixed the issue. I don't know how but it did. It would be helpful if you can maybe explain how this happened lol, maybe someone else might find something useful if they get into this issue (I really hope they don't)
Thanks!

**Reply by cromewar** [reply]
Nice to know your problem is solved now, feel free to make any other question.
